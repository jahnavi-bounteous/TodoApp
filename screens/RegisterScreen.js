import React, { useState } from "react";
import { View, Text, StyleSheet, TextInput, TouchableOpacity, Alert } from 'react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { useNavigation } from '@react-navigation/native';
import { MaterialIcons } from '@expo/vector-icons';

const TodoRegisterScreen = () => {
    const [username, setUsername] = useState('');
    const [firstName, setFirstName] = useState('');
    const [lastName, setLastName] = useState('');
    const [email, setEmail] = useState('');
    const [password, setPassword] = useState('');
    const [confirmPassword, setConfirmPassword] = useState('');
    const [isPasswordVisible, setIsPasswordVisible] = useState(false);
    const [isConfirmPasswordVisible, setIsConfirmPasswordVisible] = useState(false);
    const navigation = useNavigation();

    const handleRegisterPress = async () => {
        try {
            if (password !== confirmPassword) {
                Alert.alert('Error', 'Passwords do not match.');
                return;
            }
            
            // Check if user already exists
            const existingUser = await AsyncStorage.getItem('userData');
            if (existingUser) {
                const existingUserData = JSON.parse(existingUser);
                if (existingUserData.username === username) {
                    Alert.alert('Error', 'Username already exists. Please choose a different one.');
                    return;
                }
            }
            
            // Store user data in AsyncStorage
            await AsyncStorage.setItem(`userData_${username}`, JSON.stringify({ 
                username, 
                password, 
                firstName, 
                lastName, 
                email 
            }));
            // Create an empty todos array for the new user
            await AsyncStorage.setItem(`todos_${username}`, JSON.stringify([]));
            
            // Log entire data after registration
            const allKeys = await AsyncStorage.getAllKeys();
            const allData = await AsyncStorage.multiGet(allKeys);
            console.log('All data in storage:', allData);
            // setLoggedInUser(username);
            navigation.navigate('Login');
        } catch (error) {
            console.error('Error storing user data:', error);
            Alert.alert('Error', 'An error occurred while registering. Please try again.');
        }
    };  
    
    return (
        <View style={styles.container}>
            <TextInput
                style={styles.input}
                placeholder="Enter Username"
                value={username}
                onChangeText={text => setUsername(text)}
            />
            <View style={{ flexDirection: 'row' }}>
                <TextInput
                    style={styles.input2}
                    placeholder="Firstname"
                    value={firstName}
                    onChangeText={text => setFirstName(text)}
                />
                <View style={{ width: 10 }} />
                <TextInput
                    style={styles.input2}
                    placeholder="Lastname"
                    value={lastName}
                    onChangeText={text => setLastName(text)}
                />
            </View>
            <TextInput
                style={styles.input}
                placeholder="EmailId"
                value={email}
                onChangeText={text => setEmail(text)}
            />
            <View style={styles.passwordContainer}>
                <TextInput
                    style={styles.passwordInput}
                    placeholder="Enter Password"
                    secureTextEntry={!isPasswordVisible}
                    value={password}
                    onChangeText={text => setPassword(text)}
                />
                <TouchableOpacity
                    onPress={() => setIsPasswordVisible(!isPasswordVisible)}
                >
                    <MaterialIcons name={isPasswordVisible ? 'visibility' : 'visibility-off'} size={24} color="black" />
                </TouchableOpacity>
            </View>
            <View style={styles.passwordContainer}>
                <TextInput
                    style={styles.passwordInput}
                    placeholder="Confirm Password"
                    secureTextEntry={!isConfirmPasswordVisible}
                    value={confirmPassword}
                    onChangeText={text => setConfirmPassword(text)}
                />
                <TouchableOpacity
                    onPress={() => setIsConfirmPasswordVisible(!isConfirmPasswordVisible)}
                >
                    <MaterialIcons name={isConfirmPasswordVisible ? 'visibility' : 'visibility-off'} size={24} color="black" />
                </TouchableOpacity>
            </View>
            <TouchableOpacity
                style={[styles.textButton, { backgroundColor: "#1e90ff" }]}
                onPress={handleRegisterPress}
            >
                <Text style={styles.buttonText}>
                    Register
                </Text>
            </TouchableOpacity>
            <View style={{ flexDirection: 'row' }}>
                <Text>
                    Already have an account ?
                </Text>
                <Text> </Text>
                <TouchableOpacity onPress={() => navigation.navigate('Login')}>
                    <Text style={{ color: 'blue', textDecorationLine: 'underline' }}>
                        Login
                    </Text>
                </TouchableOpacity>
            </View>
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
    },
    input: {
        width: '80%',
        borderColor: '#1e90ff',
        borderWidth: 2,
        borderRadius: 10,
        fontSize: 16,
        padding: 16,
        marginBottom: 10
    },
    input2: {
        width: '39%',
        borderColor: '#1e90ff',
        borderWidth: 2,
        borderRadius: 10,
        fontSize: 16,
        padding: 16,
        marginBottom: 10
    },
    passwordContainer: {
        flexDirection: 'row',
        alignItems: 'center',
        width: '80%',
        borderColor: '#1e90ff',
        borderWidth: 2,
        borderRadius: 10,
        marginBottom: 10,
    },
    passwordInput: {
        flex: 1,
        fontSize: 16,
        padding: 16,
    },
    textButton: {
        width: '80%',
        borderRadius: 6,
        paddingVertical: 12,
        marginVertical: 10,
        alignItems: "center",
        shadowColor: "#000",
        shadowOffset: { width: 0, height: 2 },
        shadowOpacity: 0.8,
        shadowRadius: 3,
    },
    buttonText: {
        color: '#fff',
        fontSize: 16,
    }
})

export default TodoRegisterScreen;
