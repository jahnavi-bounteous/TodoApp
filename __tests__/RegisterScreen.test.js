import React from 'react';
import renderer from 'react-test-renderer';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { TouchableOpacity, TextInput } from 'react-native';
import RegisterScreen from '../screens/RegisterScreen';
import { NavigationContainer } from '@react-navigation/native';

// Mock AsyncStorage functions
jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(),
  setItem: jest.fn(),
}));

// Mock NavigationContainer
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => ({ navigate: jest.fn() }),
}));

// Mock BackHandler
jest.mock('react-native', () => {
  const ActualReactNative = jest.requireActual('react-native');
  return {
    ...ActualReactNative,
    BackHandler: {
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
    },
  };
});

jest.mock('react-native/Libraries/Settings/Settings', () => {
  const SettingsManager = {
    settings: {
      RNDeviceInfo: {
        uniqueId: 'unique-id',
      },
    },
  };
  return SettingsManager;
});

describe('<TodoRegisterScreen />', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders correctly', () => {
    const tree = renderer
      .create(
        <NavigationContainer>
          <RegisterScreen />
        </NavigationContainer>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

//   test('registers successfully with correct credentials', async () => {
//     AsyncStorage.getItem.mockResolvedValueOnce(null);
//     AsyncStorage.setItem.mockResolvedValueOnce(null);
//     const navigate = jest.fn();
//     const tree = renderer
//       .create(
//         <NavigationContainer>
//           <RegisterScreen />
//         </NavigationContainer>
//       )
//       .getInstance();
//     tree.handleRegisterPress();
//     await Promise.resolve();
//     expect(navigate).toHaveBeenCalledWith('Login');
//   });

//   test('displays an error message when passwords do not match', async () => {
//     const tree = renderer
//       .create(
//         <NavigationContainer>
//           <RegisterScreen />
//         </NavigationContainer>
//       )
//       .getInstance();
//     tree.setState({
//       username: 'test',
//       password: 'password',
//       confirmPassword: 'password1',
//     });
//     tree.handleRegisterPress();
//     expect(tree.state.error).toEqual('Passwords do not match.');
//   });

//   test('displays an error message when username already exists', async () => {
//     AsyncStorage.getItem.mockResolvedValueOnce(
//       JSON.stringify({ username: 'existingUser' })
//     );
//     const tree = renderer
//       .create(
//         <NavigationContainer>
//           <RegisterScreen />
//         </NavigationContainer>
//       )
//       .getInstance();
//     tree.setState({
//       username: 'existingUser',
//       password: 'password',
//       confirmPassword: 'password',
//     });
//     tree.handleRegisterPress();
//     expect(tree.state.error).toEqual('Username already exists. Please choose a different one.');
//   });

//   test('displays an error message when AsyncStorage throws an error', async () => {
//     AsyncStorage.getItem.mockRejectedValueOnce('AsyncStorage error');
//     const tree = renderer
//       .create(
//         <NavigationContainer>
//           <RegisterScreen />
//         </NavigationContainer>
//       )
//       .getInstance();
//     tree.setState({
//       username: 'test',
//       password: 'password',
//       confirmPassword: 'password',
//     });
//     tree.handleRegisterPress();
//     expect(tree.state.error).toEqual('An error occurred while registering. Please try again.');
//   });
});
