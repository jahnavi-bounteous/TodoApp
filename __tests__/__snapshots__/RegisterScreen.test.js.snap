// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterScreen /> renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <TextInput
    onChangeText={[Function]}
    placeholder="Enter Username"
    style={
      {
        "borderColor": "#1e90ff",
        "borderRadius": 10,
        "borderWidth": 2,
        "fontSize": 16,
        "marginBottom": 10,
        "padding": 16,
        "width": "80%",
      }
    }
    value=""
  />
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Firstname"
      style={
        {
          "borderColor": "#1e90ff",
          "borderRadius": 10,
          "borderWidth": 2,
          "fontSize": 16,
          "marginBottom": 10,
          "padding": 16,
          "width": "39%",
        }
      }
      value=""
    />
    <View
      style={
        {
          "width": 10,
        }
      }
    />
    <TextInput
      onChangeText={[Function]}
      placeholder="Lastname"
      style={
        {
          "borderColor": "#1e90ff",
          "borderRadius": 10,
          "borderWidth": 2,
          "fontSize": 16,
          "marginBottom": 10,
          "padding": 16,
          "width": "39%",
        }
      }
      value=""
    />
  </View>
  <TextInput
    onChangeText={[Function]}
    placeholder="EmailId"
    style={
      {
        "borderColor": "#1e90ff",
        "borderRadius": 10,
        "borderWidth": 2,
        "fontSize": 16,
        "marginBottom": 10,
        "padding": 16,
        "width": "80%",
      }
    }
    value=""
  />
  <View
    style={
      {
        "alignItems": "center",
        "borderColor": "#1e90ff",
        "borderRadius": 10,
        "borderWidth": 2,
        "flexDirection": "row",
        "marginBottom": 10,
        "width": "80%",
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Enter Password"
      secureTextEntry={true}
      style={
        {
          "flex": 1,
          "fontSize": 16,
          "padding": 16,
        }
      }
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "borderColor": "#1e90ff",
        "borderRadius": 10,
        "borderWidth": 2,
        "flexDirection": "row",
        "marginBottom": 10,
        "width": "80%",
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Confirm Password"
      secureTextEntry={true}
      style={
        {
          "flex": 1,
          "fontSize": 16,
          "padding": 16,
        }
      }
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#1e90ff",
        "borderRadius": 6,
        "marginVertical": 10,
        "opacity": 1,
        "paddingVertical": 12,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.8,
        "shadowRadius": 3,
        "width": "80%",
      }
    }
  >
    <Text
      style={
        {
          "color": "#fff",
          "fontSize": 16,
        }
      }
    >
      Register
    </Text>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <Text>
      Already have an account ?
    </Text>
    <Text>
       
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "blue",
            "textDecorationLine": "underline",
          }
        }
      >
        Login
      </Text>
    </View>
  </View>
</View>
`;

exports[`<TodoRegisterScreen /> renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <TextInput
    onChangeText={[Function]}
    placeholder="Enter Username"
    style={
      {
        "borderColor": "#1e90ff",
        "borderRadius": 10,
        "borderWidth": 2,
        "fontSize": 16,
        "marginBottom": 10,
        "padding": 16,
        "width": "80%",
      }
    }
    value=""
  />
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Firstname"
      style={
        {
          "borderColor": "#1e90ff",
          "borderRadius": 10,
          "borderWidth": 2,
          "fontSize": 16,
          "marginBottom": 10,
          "padding": 16,
          "width": "39%",
        }
      }
      value=""
    />
    <View
      style={
        {
          "width": 10,
        }
      }
    />
    <TextInput
      onChangeText={[Function]}
      placeholder="Lastname"
      style={
        {
          "borderColor": "#1e90ff",
          "borderRadius": 10,
          "borderWidth": 2,
          "fontSize": 16,
          "marginBottom": 10,
          "padding": 16,
          "width": "39%",
        }
      }
      value=""
    />
  </View>
  <TextInput
    onChangeText={[Function]}
    placeholder="EmailId"
    style={
      {
        "borderColor": "#1e90ff",
        "borderRadius": 10,
        "borderWidth": 2,
        "fontSize": 16,
        "marginBottom": 10,
        "padding": 16,
        "width": "80%",
      }
    }
    value=""
  />
  <View
    style={
      {
        "alignItems": "center",
        "borderColor": "#1e90ff",
        "borderRadius": 10,
        "borderWidth": 2,
        "flexDirection": "row",
        "marginBottom": 10,
        "width": "80%",
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Enter Password"
      secureTextEntry={true}
      style={
        {
          "flex": 1,
          "fontSize": 16,
          "padding": 16,
        }
      }
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "borderColor": "#1e90ff",
        "borderRadius": 10,
        "borderWidth": 2,
        "flexDirection": "row",
        "marginBottom": 10,
        "width": "80%",
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Confirm Password"
      secureTextEntry={true}
      style={
        {
          "flex": 1,
          "fontSize": 16,
          "padding": 16,
        }
      }
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#1e90ff",
        "borderRadius": 6,
        "marginVertical": 10,
        "opacity": 1,
        "paddingVertical": 12,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.8,
        "shadowRadius": 3,
        "width": "80%",
      }
    }
  >
    <Text
      style={
        {
          "color": "#fff",
          "fontSize": 16,
        }
      }
    >
      Register
    </Text>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <Text>
      Already have an account ?
    </Text>
    <Text>
       
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "blue",
            "textDecorationLine": "underline",
          }
        }
      >
        Login
      </Text>
    </View>
  </View>
</View>
`;
